Add decisiontransformer to onnx config

skanjila opened the pull request:
# What does this PR do?
Adds an onnx configuration for decision transformer


Fixes # (issue)
https://github.com/huggingface/transformers/issues/18191

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?
In progress while PR is being reviewed

## Who can review?
@ChainYo @regisss  Here's a first cut, I will be reading the documentation on testing this while the PR is being reviewed, again appreciate any help you can provide as this is my first time through this process

skanjila committed:
initial commit

skanjila committed:
Merge remote-tracking branch 'upstream/main' into add-decisiontransformer-to-onnx

skanjila committed:
added onnx config for decision transformer

skanjila committed:
some typo fixes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19189). All of your documentation changes will be reflected on that endpoint.

skanjila commented:
@ChainYo @regisss bumping this PR

ChainYo commented on src/transformers/models/decision_transformer/configuration_decision_transformer.py:
```diff
@@ -79,7 +81,7 @@ class DecisionTransformerConfig(PretrainedConfig):
         scale_attn_weights (`bool`, *optional*, defaults to `True`):
             Scale attention weights by dividing by sqrt(hidden_size)..
-        use_cache (`bool`, *optional*, defaults to `True`):
+        use_cache (`bool`, *optiofrom typing import Mappingnal*, defaults to `True`):
```
It looks like a random copy/paste text here :)

skanjila replied:
fixed

ChainYo commented:
> @ChainYo @regisss bumping this PR

It looks good to me!
Thanks for iterating on this, @skanjila.

regisss reviewed (changes requested):
Thanks for this PR @skanjila! It looks good to me, just a few additional things to add to @ChainYo's comment :)

You should add an entry for Decision Transformer [here](https://github.com/huggingface/transformers/blob/5cd16f01db3b5499d4665e8624801ed30ba87bdd/tests/onnx/test_onnx_v2.py#L182) so that it is tested within the CI. For the architecture to test, you can just give the smallest one.

It seems some tests failed because of the code format, please run `make style` from the Transformers' root repository.

skanjila committed:
Merge remote-tracking branch 'upstream/main' into add-decisiontransformer-to-onnx

skanjila commented:
@ChainYo @regisss  I reran make style and everything passed locally and also added an entry to DecisionTransformer in the test_onnx_v2.py file, please have a look now and lmk what else is missing

regisss commented:
@skanjila Hmm too many files got reformatted, I think your version of Black is different from what we use. Could you run `pip install transformers["quality"]` and then `make style`?

skanjila commented:
@regisss done, and make style worked, please recheck

regisss commented:
@skanjila You also need to run `make fix-copies`, I forgot this one my bad

And there are still many reformatted files, did you remove the files from the previous commit?

skanjila commented:
@regisss here is what i did:
1) to get past the reformatting of too many files I reverted back to a previous commit
2) I added decision_transformer to test_onnx_v2.py
3) I ran make style successfully after installing transformers["quality"]
4) I ran make fix-copies successfully

Hopefully this will be it, let me know your thoughts on anything else I may have missed

skanjila committed:
make commands rerun and mass formatting reverted

skanjila commented:
@regisss here is what I did
1) I reset git to an earlier commit before all my changes
2) I added back the entry for the decision_tranformer in test_onnx_v2.py
3) I reran make style after installing transformers["quality"]
4) I ran make fix-copies

Should be good to go, please have a look , I look forward to finishing this work.

regisss commented on tests/onnx/test_onnx_v2.py:
```diff
@@ -191,6 +191,7 @@ def test_values_override(self):
     ("deberta", "microsoft/deberta-base"),
     ("deberta-v2", "microsoft/deberta-v2-xlarge"),
     ("convnext", "facebook/convnext-tiny-224"),
+    ("decision_transformer", "huggingface/decision_transformer"),
```
```suggestion
    ("decision-transformer", "edbeeching/decision-transformer-gym-hopper-medium"),
```
It seems this checkpoint does not exist in the Hugging Face Hub. I suggest this one.

skanjila replied:
@regisss done, I modified test_onnx_v2 to point to the correct decision-transformer model as you identified above, also listed below is the output of the command you specified

saikatkanjilal@system76-pc:~/code/opensource/transformers$ RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "decision-transformer"
========================================================================================= test session starts =========================================================================================
platform linux -- Python 3.10.0rc1, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/saikatkanjilal/code/opensource/transformers, configfile: setup.cfg
collected 432 items / 432 deselected / 0 selected                                                                                                                                                     

========================================================================================== warnings summary ===========================================================================================
src/transformers/testing_utils.py:28
  /home/saikatkanjilal/code/opensource/transformers/src/transformers/testing_utils.py:28: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.util import strtobool

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================= 432 deselected, 1 warning in 0.30s ==================================================================================


As you can see I just received warnings but no errors, are we close?

regisss replied:
@skanjila The test was not run because it could not find it. It happens because in `test_onnx_v2.py` you did not change the first element of the tuple to `decision-transformer` (`-` instead of `_` is a convention we use for model names). Could you change this and re-run the test command please?

skanjila replied:
@regisss I get this error when I put dash instead of underscore:  E   KeyError: "decision-transformer is not supported yet. Only ['albert', 'bart', 'beit', 'bert', 'big-bird', 'bigbird-pegasus', 'blenderbot', 'blenderbot-small', 'bloom', 'camembert', 'clip', 'codegen', 'convbert', 'convnext', 'data2vec-text', 'data2vec-vision', 'deberta', 'deberta-v2', 'decision_transformer', 'deit', 'detr', 'distilbert', 'electra', 'flaubert', 'gpt2', 'gptj', 'gpt-neo', 'groupvit', 'ibert', 'layoutlm', 'layoutlmv3', 'levit', 'longt5', 'longformer', 'marian', 'mbart', 'mobilebert', 'mobilevit', 'mt5', 'm2m-100', 'owlvit', 'perceiver', 'resnet', 'roberta', 'roformer', 'segformer', 'squeezebert', 't5', 'vit', 'xlm', 'xlm-roberta', 'yolos'] are supported. If you want to support decision-transformer please propose a PR or open up an issue."


Note that the directory for decision transformer under model has an underscore

skanjila replied:
@regisss if you look at the error message it's actually looking for an underscore, did I miss something here?

regisss replied:
@skanjila  Ah you also need to put the dash in the `features.py` file, I missed that

skanjila replied:
@regisss I added the dash in features.py as shown here and in the latest commit :  "decision-transformer": supported_features_mapping(
            "default",
            onnx_config_cls="models.decision_transformer.DecisionTransformerOnnxConfig",
        ),  , listed below are the results:
saikatkanjilal@system76-pc:~/code/opensource/transformers$ RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "decision-transformer"
========================================================================================= test session starts =========================================================================================
platform linux -- Python 3.10.0rc1, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/saikatkanjilal/code/opensource/transformers, configfile: setup.cfg
collected 432 items / 432 deselected / 0 selected                                                                                                                                                     

========================================================================================== warnings summary ===========================================================================================
src/transformers/testing_utils.py:28
  /home/saikatkanjilal/code/opensource/transformers/src/transformers/testing_utils.py:28: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.util import strtobool

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================= 432 deselected, 1 warning in 0.30s ==================================================================================


Let me know next steps

regisss replied:
@skanjila Sorry it should be an underscore in the pytest command actually: `RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "decision_transformer"`.

I just ran it on my laptop and it fails because the default method for generating dummy inputs to export the model to the ONNX format expects a preprocessor but got none. Which makes sense according to the configuration of the model.
I don't know this model at all but I think you will have to override the [`generate_dummy_inputs`](https://github.com/huggingface/transformers/blob/c875a96eb154f2ea4a831a48b47c37c1c3da99e7/src/transformers/onnx/config.py#L264) method in `DecisionTransformerOnnxConfig` to adapt it for this model. According to the [documentation](https://huggingface.co/docs/transformers/v4.22.2/en/model_doc/decision_transformer#transformers.DecisionTransformerModel.forward), the inputs to the `forward` method of the model are states, actions and rewards. So you will have to generate dummy states, dummy actions and dummy rewards.

skanjila replied:
@regisss sigh sorry I feel like we're lost about 10 steps of progress based on the above work that still needs to be done, the above is going to take a bit of time, let me take a crack at this and I will make the changes to the PR and I may need help on the above

regisss replied:
@skanjila  Yeah I know I didn't anticipate this since I've never used this model :confused: 

I can definitely help with this. Have you already used this model within the Transformers library? Because if you're familiar with the inputs it takes it could be quite straightforward.

skanjila replied:
@regisss no I have not, give me a bit of time to learn this, thats the whole point of me contributing anyway, i will ping you through the journey

regisss reviewed (changes requested):
Thanks @skanjila, we are almost there! I wrote one last comment to change the name of the checkpoint used for testing. Could you run this test locally with `RUN_SLOW=1 pytest tests/onnx/test_onnx_v2.py -k "decision-transformer"` and let me know if it passes please?

skanjila committed:
added the correct decision transformer config into the tests

skanjila committed:
Merge remote-tracking branch 'upstream/main' into add-decisiontransformer-to-onnx

skanjila committed:
Merge remote-tracking branch 'upstream/main' into add-decisiontransformer-to-onnx

skanjila committed:
Merge remote-tracking branch 'upstream/main' into add-decisiontransformer-to-onnx

skanjila committed:
changed underscore to a dash in features.py
